LEGO1/mxstring.cpp

```diff
@@ -0,0 +1,30 @@
+#include "mxstring.h"
+#include <stdlib.h>
+#include <string.h>
+
+MxString::MxString()
+{
+  // Set string to one char in length and set that char to null terminator
+  this->m_data = (char *)malloc(1);
+  this->m_data[0] = 0;
+  this->m_length = 0;
+}
+
+// TODO: this *mostly* matches, again weird with the comparison
+const MxString &MxString::operator=(const char *param)
+{
+  if (this->m_data != param)
+  {
+    free(this->m_data);
+    this->m_length = strlen(param);
+    this->m_data = (char *)malloc(this->m_length + 1);
+    strcpy(this->m_data, param);

```

Comment:
I think you need to do `memcpy` here, otherwise the `strcpy` might (will) do another `strlen`.
```suggestion
    memcpy(this->m_data, param, this->m_length + 1);
```

Reply:
@madebr Keep in mind, the goal is not to write _good_ code, it's to write _accurate_ code. This is faithful to what the game does.

Reply:
OG (=original game) calculates the length of the string only once.

Reply:
I think you're either looking at the wrong function or misinterpreting it. There are two length calculations here (the two do/whiles are both counting until the null terminator), then the actual copy is done at the very end (the two for loops).

![image](https://github.com/isledecomp/isle/assets/34096995/e4bc948c-7271-499b-b769-eee304d164c4)

Reply:
Aren't you looking at the `MxString::MxString(const char*)` constructor at `0x100ae350`?
The `operator=(const char *)` method is at `0x100ae510`.

Reply:
This is why I think adding comments containing the virtual addresses, would be useful.

Reply:
No, I'm looking at `0x100ae510`. And it's a moot point because both of them do exactly the same thing, they both have two loops to calculate the string length, one for `strlen` and one for `strcpy`.

> This is why I think adding comments containing the virtual addresses, would be useful.

Yes, I'm planning on enforcing this soon, along with the script I mention in #21 that will use those comments to verify the accuracy of each function (the comments will need to be in a specific format for this reason).

Reply:
You're right.
Thanks.